Documentation/ExtensionArchitecture/Extbase/Reference/FileUpload.rst

```diff
@@ -0,0 +1,489 @@
+
+..  literalinclude:: _FileUpload/_tx_myextension_domain_model_blog.php
+    :caption: EXT:my_extension/Configuration/TCA/tx_myextension_domain_model_blog.php
+
+Once this is set up, you can create/edit records through the TYPO3
+backend, attach a single or multiple files in it. Then using a normal
+controller and fluid template, you can display an image:
+
+..  literalinclude:: _FileUpload/_BlogController.php
+    :caption: EXT:my_extension/Classes/Controller/BlogController.php
+
+..  literalinclude:: _FileUpload/_Show.html
+    :caption: EXT:my_extension/Resources/Private/Templates/Blog/Show.html
+
+On the PHP-side within controllers, you can use the usual
+:php:`$blogItem->getSingleFile()` and :php:`$blogItem->getMultipleFilers()`
+Extbase getters to retrieve the FileReference object.
+
+..  _extbase_fileupload_writing:
+
+Writing FileReference entries
+-----------------------------
+
+..  _extbase_fileupload_writing-manual:
+
+Manual handling
+...............
+
+With TYPO3 versions below v13, attaching files to an Extbase Domain Model
+was only possible by either:
+
+*  Manually evaluate the :php:`$_FILES` data, process and validate the data,
+   use raw QueryBuilder write actions on :sql:`sys_file` and :sql:`sys_file_reference`
+   to persist the files quickly, or use at least some API methods:
+
+    ..  literalinclude:: _FileUpload/_ApiUpload.php
+        :caption: EXT:my_extension/Classes/Controller/BlogController.php, excerpt
+
+   Instead of raw access to :php:`$_FILES`, starting with TYPO3 v12 the recommendation
+   is to utilize the :ref:`UploadedFile objects instead of $_FILES <changelog:breaking-97214>`.
+   In that case, validators can be used for custom UploadedFile objects to specify restrictions
+   on file types, file sizes and image dimensions.
+
+*  Use (or better: adapt) a more complex implementation by using Extbase TypeConverters,
+   as provided by `Helmut Hummel's EXT:upload_example <https://github.com/helhum/upload_example>`__.
+   This extension is no longer maintained and will not work without larger adaptation for
+   TYPO3 v12 compatibility.
+
+..  _extbase_fileupload_writing-attributes:
+
+Automatic handling based on PHP Attributes
+..........................................
+
+Starting with TYPO3 v13 it is finally possible to streamline this with commonly
+know Extbase logic, as implemented via
+:ref:`Feature: #103511 - Introduce Extbase file upload and deletion handling <changelog:feature-103511-1711894330>`.
+
+An example implementation of this can be found on
+`Torben Hansen's EXT:extbase-upload <https://github.com/derhansen/extbase_upload>`__
+repository.
+
+The general idea is to use PHP attributes within the Extbase Model, and for the upload
+use a custom ViewHelper:
+
+..  literalinclude:: _FileUpload/_BlogEnhanced.php
+    :caption: EXT:my_extension/Classes/Domain/Model/Blog.php, using FileUpload attributes
+
+..  literalinclude:: _FileUpload/_Upload.html
```

Comment:
Would add an introduction text like: "and the Fluid template" or similar to split up two consecutive code blocks.

Reply:
done